kernel_gateway/mixins.py

```diff
@@ -64,8 +64,14 @@ def prepare(self):
         with the `@web.authenticated` decorated methods in the notebook
         package.
         """
+        request_is_options = False
+        if 'request' in dir(self):
```

Comment:
I'm not too firm in Python myself, but I think the calls to `dir()` are creating two temporary lists in this section, just to check whether there are attributes. Try the following instead (I haven't checked the exact syntax):
```python
   request_is_options = hasattr(self, 'request') and getattr(self.request, 'method', '') == 'OPTIONS'
```
https://stackoverflow.com/a/610893/5629418
https://stackoverflow.com/a/611708/5629418
This could also be inline below as
```python
if server_token and not (...):
```
But looking further below, `client_token` is also read from `self.request` without checking whether that exists, and a request without `method` wouldn't make any sense. I agree with @parente that it is probably a problem with the test setup if `self.request` doesn't exist.